### PR TITLE
[JENKINS-41124] Move name mangling support responsibility to AbstractFolder

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-41124](https://issues.jenkins-ci.org/browse/JENKINS-41124)
Downstream of https://github.com/jenkinsci/cloudbees-folder-plugin/pull/82

Additional test cases:
- [x] Need a test that verifies after migration and then a restart that all is still as expected.
- [x] Need a test that verifies after a migration and then a reload that all is still as expected.
- [x] Need a test that verifies reloading the folder and not all of Jenkins leaves all still as expected.
- [x] Need a test that verifies after a migration and thene a restart and then a reload that all is still as expected.
- [x] Need a set of all those migration tests using the migration from a 2.0.0 pristine set of items
- [x] Need a set of all those migration tests using the migration from a 1.x set of items that was migrated to 2.0.0 and then lands on this set
- [x] Need a set of migration tests that verifies a pristine instance is in the same state as a migrated instance

@reviewbybees esp @jglick @michaelneale

### CODE REVIEW

**Comment by michaelneale** [comment]
I tried the latest snapshot out in blue ocean but with no luck: 

```
[ERROR] Failed to execute goal org.kohsuke:access-modifier-checker:1.7:enforce (default-enforce) on project blueocean-events: Failed to enforce @Restricted constraints: JAR entry META-INF/annotations/org.kohsuke.accmod.Restricted not found in /Users/michaelneale/.m2/repository/org/jenkins-ci/plugins/branch-api/2.0.2-SNAPSHOT/branch-api-2.0.2-SNAPSHOT.jar -> [Help 1]
```

branch api in this case was: 2.0.2-20170119.100545-2

previously: 2.0.2-20170118.220727-1 which did work

**Comment by stephenc** [comment]
@reviewbybees this should be ready for review now

**Comment by michaelneale** [comment]
@stephenc do you have a published snap to try?

**Comment by stephenc** [comment]
@michaelneale I'll re-publish in a few minutes... just sorting out a copy of the bits that were pulled from the release UC so that JENKINS-41207 can at least not be running buggy beta versions

**Comment by stephenc** [comment]
@michaelneale `2.0.2-20170119.233611-4` should be the branch-api snapshot. `5.17-20170119.163248-6` for folders

**Comment by michaelneale** [comment]
It builds ok, URIS look good, but PRs don't quite show up right (I think they are ok but not in blue ocean). When I ran a re-index I saw this (likely irrelevant): 

```
SEVERE: Executor threw an exception
java.lang.NoSuchMethodError: hudson.plugins.git.GitSCM.setBrowser(Lhudson/plugins/git/browser/GitRepositoryBrowser;)V
	at org.jenkinsci.plugins.github_branch_source.GitHubSCMSource.build(GitHubSCMSource.java:818)
	at jenkins.scm.api.SCMSource.build(SCMSource.java:778)
	at jenkins.branch.MultiBranchProject.newBranch(MultiBranchProject.java:400)
```

I would need to ask @vivek for help diagnosing why PRs don't look right now.

**Comment by stephenc** [comment]
No such method indicates git plugin is not 2.6.2 but 3.0.0 or 3.0.1 (you want 3.0.2 or 3.0.3)

**Comment by stephenc** [comment]
That No Such Method error will cause some branches to fail to index correctly IIRC

**Comment by stephenc** [comment]
Created [JENKINS-41244](https://issues.jenkins-ci.org/browse/JENKINS-41244) to track

File: src/main/java/jenkins/branch/MultiBranchProjectFactory.java [hunk 1]
```diff
@@ -205,7 +205,7 @@ public MultiBranchProjectFactoryDescriptor getDescriptor() {
          * @param source a repository
          * @return criteria for treating its branches as a match
          */
-        @NonNull
+        @CheckForNull
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProjectFactory.java]: How on earth is this related? And the Javadoc does not describe what a null return value would mean.

File: src/main/java/jenkins/branch/NameEncoder.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.branch;
+
+/**
+ * Encodes names that are not nice so that they are safe to use as url path segments.
+ * We don't want to do a full url encoding, only replace problematic names with {@code %} escaped variants so
+ * that when double encoded by Stapler etc we bypass issues.
```
Comment (jglick) [inline src/main/java/jenkins/branch/NameEncoder.java]: `etc.`

File: src/main/java/jenkins/branch/NameEncoder.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.branch;
+
+/**
+ * Encodes names that are not nice so that they are safe to use as url path segments.
```
Comment (jglick) [inline src/main/java/jenkins/branch/NameEncoder.java]: URL

File: src/main/java/jenkins/branch/NameEncoder.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    private NameEncoder() {
+        throw new IllegalAccessError("Utility class");
+    }
+
+    public static String encode(String name) {
+        if ("".equals(name)) {
+            return "%00";
+        }
+        if (".".equals(name)) {
+            // just enough escaping to bypass the URL segment meaning of ".."
```
Comment (jglick) [inline src/main/java/jenkins/branch/NameEncoder.java]: `.` I suppose you mean.

File: src/main/java/jenkins/branch/NameEncoder.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+            return "%2E";
+        }
+        if ("..".equals(name)) {
+            // just enough escaping to bypass the URL segment meaning of ".."
+            return "%2E.";
```
Comment (jglick) [inline src/main/java/jenkins/branch/NameEncoder.java]: I wonder if some web servers will treat an attempt to visit such URLs as an attempted hack and block them. I know we have had problems whereby valid URLs like `http://jenkins/descriptorById/whatever/thing/foo%2Fbar/` correctly call `Whatever.DescriptorImpl.getThing("foo/bar")` in most cases but in some configurations of Tomcat produce a 404 for supposed security reasons.

File: src/main/java/jenkins/branch/NameEncoder.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                    break;
+                case '%':
+                    buf.append("%25");
+                    break;
+                case '/':
+                    buf.append("%2F");
+                    break;
+                case '?':
+                buf.append("%3F");
+                break;
```
Comment (jglick) [inline src/main/java/jenkins/branch/NameEncoder.java]: indentation

File: src/main/java/jenkins/branch/NameEncoder.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        if ("%00".equals(name)) {
+            return "";
+        }
+        if ("%2E".equals(name)) {
+            return ".";
+        }
+        if ("%2E.".equals(name)) {
+            return "..";
+        }
+        if (name.indexOf('%') == -1) {
```
Comment (jglick) [inline src/main/java/jenkins/branch/NameEncoder.java]: move this normal case to the top for efficiency

File: src/main/java/jenkins/branch/NameEncoder.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            return name;
+        }
+        StringBuilder buf = new StringBuilder(name.length() + 16);
+        int matchIndex = 0;
+        char match = 0;
+        for (char c : name.toCharArray()) {
+            switch (matchIndex) {
+                case 1:
+                    if (c == '2' || c == '3' || c == '5') {
```
Comment (jglick) [inline src/main/java/jenkins/branch/NameEncoder.java]: Is this machine-generated code?!

File: src/test/java/integration/BrandingTest.java [hunk 9]
```diff
@@ -31,6 +31,7 @@
 import hudson.model.ListView;
 import hudson.model.Result;
 import hudson.model.TopLevelItem;
+import hudson.util.StreamTaskListener;
```
Comment (jglick) [inline src/test/java/integration/BrandingTest.java]: Why?

File: src/test/java/jenkins/branch/WorkspaceLocatorImplTest.java [hunk 10]
```diff
@@ -74,11 +74,11 @@ public void locate() throws Exception {
         stuff.scheduleBuild2(0).getFuture().get();
         r.waitUntilNoActivity();
         showComputation(stuff);
-        FreeStyleProject master = r.jenkins.getItemByFullName("stuff/dev-flow.bn29c1", FreeStyleProject.class);
+        FreeStyleProject master = r.jenkins.getItemByFullName("stuff/dev%2fflow", FreeStyleProject.class);
```
Comment (jglick) [inline src/test/java/jenkins/branch/WorkspaceLocatorImplTest.java]: `%2F` or `%2f`?

File: src/test/java/integration/MigrationTest.java [hunk 11]
```diff
@@ -78,57 +86,273 @@ public static void closeSCM() {
...
+    /**
+     * Checks that data migrated from 1.x to 2.0.0 name mangling is still valid when re-migrated to 2.0.2 name mangling
+     */
+    @Test
+    @LocalData
+    public void nameMangling_2_folder_reload() throws Exception {
+        r.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                OrganizationFolder foo = (OrganizationFolder) r.j.jenkins.getItem("foo");
```
Comment (jglick) [inline src/test/java/integration/MigrationTest.java]: `getItemByFullName` takes a type param

File: src/main/java/jenkins/branch/NameEncoder.java (hunk 8 above)
Comment (stephenc) [inline src/main/java/jenkins/branch/NameEncoder.java]: Are you calling me a machine?!

File: src/main/java/jenkins/branch/MultiBranchProjectFactory.java [hunk 12]
```diff
@@ -203,7 +203,7 @@ public MultiBranchProjectFactoryDescriptor getDescriptor() {
          * Defines how to decide whether or not a given repository should host our type of project.
          *
          * @param source a repository
-         * @return criteria for treating its branches as a match
+         * @return criteria for treating its branches as a match or ({@code null} if all branches match)
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProjectFactory.java]: misplaced lparen
